runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/WindmillStreamPoolSender.java

```diff
@@ -0,0 +1,159 @@
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      WorkCommitter workCommitter,
+      GetDataClient getDataClient,
+      HeartbeatSender heartbeatSender,
+      StreamingWorkScheduler streamingWorkScheduler,
+      Runnable waitForResources,
+      Function<String, Optional<ComputationState>> computationStateFetcher) {
+    this.connection = connection;
+    this.getWorkRequest = getWorkRequest;
+    this.getWorkBudget = getWorkBudget;
+    this.streamingEngineStreamFactory = streamingEngineStreamFactory;
+    this.getDataClient = getDataClient;
+    this.heartbeatSender = heartbeatSender;
+    this.streamingWorkScheduler = streamingWorkScheduler;
+    this.waitForResources = waitForResources;
+    this.computationStateFetcher = computationStateFetcher;
+    this.workCommitter = workCommitter;
+  }
+
+  public static WindmillStreamPoolSender create(
+      WindmillConnection connection,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudget getWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      WorkCommitter workCommitter,
+      GetDataClient getDataClient,
+      HeartbeatSender heartbeatSender,
+      StreamingWorkScheduler streamingWorkScheduler,
+      Runnable waitForResources,
+      Function<String, Optional<ComputationState>> computationStateFetcher) {
+    WindmillStreamPoolSender windmillStreamPoolSender =
+        new WindmillStreamPoolSender(
+            connection,
+            getWorkRequest,
+            new AtomicReference<>(getWorkBudget),
+            streamingEngineStreamFactory,
+            workCommitter,
+            getDataClient,
+            heartbeatSender,
+            streamingWorkScheduler,
+            waitForResources,
+            computationStateFetcher);
+    windmillStreamPoolSender.singleSourceWorkerHarness =
+        SingleSourceWorkerHarness.builder()
+            .setWorkCommitter(workCommitter)
+            .setGetDataClient(getDataClient)
+            .setHeartbeatSender(heartbeatSender)
+            .setStreamingWorkScheduler(streamingWorkScheduler)
+            .setWaitForResources(waitForResources)
+            .setComputationStateFetcher(computationStateFetcher)
+            .setGetWorkSender(
+                SingleSourceWorkerHarness.GetWorkSender.forStreamingEngine(
+                    windmillStreamPoolSender::createGetWorkStream))
+            .build();
+    return windmillStreamPoolSender;
+  }
+
+  private GetWorkStream createGetWorkStream(WorkItemReceiver workItemReceiver) {
+    // Creating a local variable first instead of directly assigning it to the instance variable
+    // to keep the compiler happy which otherwise complains NonNull value is being assinged to
+    // Nullable.
+    GetWorkStream workStream =
+        streamingEngineStreamFactory.createGetWorkStream(
+            connection.currentStub(), getWorkRequest, workItemReceiver);
+    workStream.start();
+    this.getWorkStream = workStream;
```

Comment:
this seems racy, it's not synchronized and we're passing this method to some other class to call which could be simultaneous with it's usage in setBudget